contracts/metatx/MinimalForwarder.sol

```diff
@@ -23,59 +28,85 @@ contract MinimalForwarder is EIP712 {
         uint256 value;
         uint256 gas;
         uint256 nonce;
+        uint48 deadline;
         bytes data;
     }
 
-    bytes32 private constant _TYPEHASH =
-        keccak256("ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,bytes data)");
-
-    mapping(address => uint256) private _nonces;
+    bytes32 private constant _FORWARD_REQUEST_TYPEHASH =
+        keccak256(
+            "ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,uint48 deadline,bytes data)"
+        );
 
     /**
      * @dev The request `from` doesn't match with the recovered `signer`.
      */
     error MinimalForwarderInvalidSigner(address signer, address from);
 
     /**
-     * @dev The request nonce doesn't match with the `current` nonce for the request signer.
+     * @dev The request `value` doesn't match with the `msg.value`.
+     */
+    error MinimalForwaderMismatchedValue(uint256 msgValue, uint256 value);
+
+    /**
+     * @dev The list of requests length doesn't match with the list of signatures length.
      */
-    error MinimalForwarderInvalidNonce(address signer, uint256 current);
+    error MinimalForwaderInvalidBatchLength(uint256 requestsLength, uint256 signaturesLength);
 
-    constructor() EIP712("MinimalForwarder", "0.0.1") {}
+    /**
+     * @dev The request `deadline` has expired.
+     */
+    error MinimalForwarderExpiredRequest(uint256 deadline);
 
-    function getNonce(address from) public view returns (uint256) {
-        return _nonces[from];
-    }
+    /**
+     * @dev See {EIP7712-constructor}.
+     */
+    constructor(string memory name, string memory version) EIP712(name, version) {}
 
-    function verify(ForwardRequest calldata req, bytes calldata signature) public view returns (bool) {
-        address signer = _recover(req, signature);
-        (bool correctFrom, bool correctNonce) = _validateReq(req, signer);
-        return correctFrom && correctNonce;
+    /**
+     * @dev Returns `true` if a request is valid for a provided `signature` at the current block.
+     */
+    function verify(ForwardRequest calldata request, bytes calldata signature) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, bool nonceMatch) = _validate(request, signature);
+        return alive && signerMatch && nonceMatch;
     }
 
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer.
+     */
     function execute(
-        ForwardRequest calldata req,
+        ForwardRequest calldata request,
         bytes calldata signature
-    ) public payable returns (bool, bytes memory) {
-        address signer = _recover(req, signature);
-        (bool correctFrom, bool correctNonce) = _validateReq(req, signer);
+    ) public payable virtual returns (bool, bytes memory) {
+        if (request.deadline < block.number) {
```

Comment:
The formula for deadline validity is written twice, once in `execute` and once in `_validateAt`. Those two formulas are in sync now, but I'd prefer if it was written in only one place to avoid the possibility for off-by-one errors.

Reply:
Not sure if it's a good idea.

The validation consists of three checks:

1. Deadline
2. Signer
3. Nonce

We can keep the same deadline check in a single shared function, but that's not the case for the other two because:

- In case the signer fails, we need to throw an error with the signer. This means that a single shared function for both signature checks should share the signer as well (or we have to recover twice unnecessarily)
- I think it's reasonable to use `_useCheckedNonce`, but that already performs an SLOAD for the nonce. Using a single shared place would require a duplicated SLOAD

We can still have a deadline check function though, but I think it's better to fuzz the `verify` and `execute` functions to check they both work the same. What do you think?

Reply:
I would rather address this at the code level than with fuzzing or testing.

The concerns you shared are valid though. We could have the check deadline function.

Reply:
> I would rather address this at the code level than with fuzzing or testing.

Same but I don't see an easy way now that we are even checking different things in the batch version vs the single version. We're not checking nonce correctness in the batch version so now the checks are even more distributed across different functions

Reply:
Ok I did this in `_validate`:

```solidity
function _validate(
    ForwardRequest calldata request,
    bytes calldata signature
) internal view virtual returns (bool alive, bool signerMatch, bool nonceMatch, address signer) {
    signer = _recoverForwardRequestSigner(request, signature);
    return (
        request.deadline >= block.number,
        signer == request.from,
        nonces(request.from) == request.nonce,
        signer
    );
}
```

Let me know if you like it, I think it looks pretty good tho

Reply:
The new `_validate` looks pretty good. The nonces are nicely handled.